src/librustc_feature/active.rs

```diff
@@ -44,28 +32,43 @@ macro_rules! declare_features {
             pub declared_lib_features: Vec<(Symbol, Span)>,
             $(
                 $(#[doc = $doc])*
-                pub $feature: bool
+                $feature: bool
             ),+
         }
 
         impl Features {
+            /// Is the given feature enabled?
+            ///
+            /// Panics if the symbol doesn't correspond to a declared feature.
+            pub fn on(&self, name: Symbol) -> bool {
+                match name {
+                    $(sym::$feature => self.$feature,)+
```

Comment:
`match` is a pretty inefficient tool in this case (in terms of performance).
Simple linear (not even binary) search in an array could be better here.

Reply:
I wish we had field pointers (like `&MyStruct::my_field`).
They can be emulated with `offset_of`, but that would mean macros again.

Reply:
Perhaps the field approach can be kept with the use of macros being minimized rather than fully eliminated.

Reply:
Initially I had a `HashSet` but was hoping that the `match` would turn into a huge lookup table and be better that way. I can certainly try a `HashSet` again and see if it changes anything. 

Otherwise, perhaps field accesses outside `check.rs` would be workable (ostensibly that's where the regressions are coming from in some hot loop somewhere?) -- that should still leave us with no macros.

Reply:
It is unclear to me -- why is the macro itself bad? It looks like this is still using a macro. It looks also like we're still generating the fields, so we why can't expose those is not clear to me.